### PR TITLE
Codegen refactoring and tests

### PR DESCRIPTION
**Pull request by StefanoFioravanzo**
This PR contains 

- a refactoring of the codegen module, basically splitting up the previously monolithic function into multiple smaller and testable functions.
- some polishing of the Jinja templates
- improvements in notebook metadata handling from CLI args to code generation
- fixes here and there to make the gen features work more organically
- unit tests for gen module

### CODE REVIEW

File: kale/codegen/generate_code.py [hunk 1]
```diff
@@ -135,19 +135,21 @@ def generate_lightweight_component(template, step_name, step_data, nb_path,
...
+    fn_code = template.render(
         step_name=step_name,
         function_body=[step_source],
         in_variables=step_marshal_in,
         out_variables=step_marshal_out,
         nb_path=nb_path,
         **metadata
     )
+    # fix code style using pep8 guidelines
+    return autopep8.fix_code(fn_code)
```
Comment (elikatsis) [inline kale/codegen/generate_code.py]: nit [but no strong opinion]:
```python
    return autopep8.fix_code(template.render(step_name=step_name,               
                                             function_body=[step_source],       
                                             in_variables=step_marshal_in,      
                                             out_variables=step_marshal_out,    
                                             nb_path=nb_path,                   
                                             **metadata)
```

File: kale/codegen/generate_code.py [hunk 2]
```diff
@@ -160,7 +162,8 @@ def generate_pipeline(template, nb_graph, step_names, lightweight_components,
         leaf_steps=leaf_steps,
         **metadata
     )
-    return pipeline_code
+    # fix code style using pep8 guidelines
+    return autopep8.fix_code(pipeline_code)
```
Comment (elikatsis) [inline kale/codegen/generate_code.py]: ditto:
```python
    return autopep8.fix_code(template.render(                                   
        lightweight_components=lightweight_components,                          
        step_names=step_names,                                                  
        step_prevs=pipeline_dependencies_tasks(nb_graph),                       
        leaf_steps=leaf_steps,                                                  
        **metadata)
```

File: kale/codegen/generate_code.py [hunk 3]
```diff
@@ -134,25 +134,21 @@ def pipeline_dependencies_tasks(g):
...
-                                   step_data, function_args, marshal_path,
-                                   auto_snapshot, nb_path):
+def generate_lightweight_component(template, step_name, step_data, nb_path,
+                                   metadata):
     """Use the function template to generate Python code."""
     step_source = step_data['source']
     step_marshal_in = step_data['ins']
     step_marshal_out = step_data['outs']
 
     # TODO: Remove some parameters and pass them with **metadata
```
Comment (elikatsis) [inline kale/codegen/generate_code.py]: nit: remove comment

File: kale/codegen/generate_code.py [hunk 4]
```diff
@@ -169,17 +165,7 @@ def generate_pipeline(template, nb_graph, step_names, lightweight_components,
...
-        pipeline_description=metadata.get('pipeline_description', ''),
-        pipeline_args=metadata.get('pipeline_args'),
-        pipeline_args_names=metadata.get('pipeline_args_names'),
-        docker_image=metadata.get('docker_image', ''),
-        volumes=metadata.get('volumes'),
-        abs_working_dir=metadata.get('abs_working_dir', None),
-        marshal_volume=metadata.get('marshal_volume'),
-        marshal_path=metadata.get('marshal_path'),
-        auto_snapshot=metadata.get('auto_snapshot')
+        **metadata
```
Comment (elikatsis) [inline kale/codegen/generate_code.py]: nit: remove `TODO` comment in line 162 (I can't put a comment on that line)

File: kale/core.py [hunk 5]
```diff
@@ -152,6 +153,22 @@ def validate_metadata(self):
                         "Provide a valid snapshot resource name if you want to"
                         " snapshot a volume. Snapshot resource name %s" %
                         k8s_name_msg)
+
+                # Convert annotations to a dictionary
+                annotations = {a['key']: a['value'] for a in
+                               v['annotations'] or []
+                               if a['key'] != '' and a['value'] != ''}
```
[the remark this answers is not in the document]
Comment (elikatsis) [reply kale/core.py]: ^ updated

File: kale/codegen/generate_code.py (hunk 1 above)
Comment (StefanoFioravanzo) [reply kale/codegen/generate_code.py]: I'd rather keep it as it is. I think it is more readable, it does not make much sense to oversimplify in this case

File: kale/codegen/generate_code.py (hunk 2 above)
Comment (StefanoFioravanzo) [reply kale/codegen/generate_code.py]: same as above

File: kale/codegen/generate_code.py [hunk 6]
```diff
@@ -3,159 +3,219 @@
...
+
+    Args:
+        g: Pipeline graph
+
+    Returns (dict): k: step_name, v: list of predecessors
+    """
+    deps = dict()
+    for step_name in nx.topological_sort(g):
+        deps[step_name] = [f"{pred}_task" for pred in
+                           g.predecessors(step_name)]
```
Comment (elikatsis) [inline kale/codegen/generate_code.py]: Python3.5 compatibility & style
```python
deps[step_name] = ["{}_task".format(pred)
                   for pred in g.predecessors(step_name)]

File: kale/codegen/generate_code.py [hunk 7]
```diff
@@ -3,159 +3,219 @@
...
                                for arg in pipeline_parameters])
-    # arguments are actually the pipeline arguments. Since we don't know precisely in which pipeline
-    # steps they are needed we just pass them to every one. The assumption is that these variables
-    # were treated as constants notebook-wise.
-    function_args = ', '.join([f"{arg}: {pipeline_parameters[arg][0]}" for arg in pipeline_parameters])
+    # Arguments are the pipeline arguments. Since we don't know precisely in
+    # what pipeline steps they are needed, we just pass them to every one.
+    # We assume there variables were not re-assigned throughout the notebook
+    function_args = ', '.join([f"{arg}: {pipeline_parameters[arg][0]}"
+                               for arg in pipeline_parameters])
```
Comment (elikatsis) [inline kale/codegen/generate_code.py]: Python3.5 compat
```python
function_args = ', '.join(["{}: {}".format(arg,
                                           pipeline_parameters[arg][0])
                           for arg in pipeline_parameters])
```

File: kale/core.py [hunk 8]
```diff
@@ -243,6 +272,18 @@ def notebook_to_graph(self):
         dependencies.dependencies_detection(pipeline_graph,
                                             ignore_symbols=to_ignore)
 
+        # add an empty step at the end of the pipeline for final snapshot
+        if self.auto_snapshot:
+            auto_snapshot_name = 'final_auto_snapshot'
+            # add a link from all the last steps of the pipeline to
+            # the final auto snapshot one.
+            leaf_steps = [x for x in pipeline_graph.nodes() if
+                          pipeline_graph.out_degree(x) == 0]
```
Comment (elikatsis) [inline kale/core.py]: nit:
```python
leaf_steps = [x for x in pipeline_graph.nodes()
              if pipeline_graph.out_degree(x) == 0]
```

File: kale/codegen/generate_code.py (hunk 7 above)
Comment (StefanoFioravanzo) [reply kale/codegen/generate_code.py]: I will leave all PY3.5 `format` refactoring to #54

**Pull request merged**